tests/pjit_test.py

```diff
@@ -166,6 +166,30 @@ def f(x):
     # Annotation from pjit
     self.assertIn("sharding={replicated}", hlo.as_hlo_text())
 
+  @with_mesh([('x', 2), ('y', 1)])
+  def testShardingConstraintPyTree(self):
+    @partial(pjit, in_axis_resources=None, out_axis_resources=None)
+    def f(x):
+      x = with_sharding_constraint(x, [P('x', 'y'), P('y', 'x')])
+      x[0]["a"] *= 2
```

Comment:
```suggestion
      x = x.copy()
      x[0]["a"] *= 2
```

Otherwise there's a side-effect on the caller's dict, right?

Reply:
Haha I'm breaking the cardinal rule of JAX! Good catch.